crates/bws/src/command/run.rs

```diff
@@ -0,0 +1,133 @@
+use std::{
+    collections::{HashMap, HashSet},
+    io::{IsTerminal, Read},
+    process,
+};
+
+use bitwarden::{
+    secrets_manager::{
+        secrets::{SecretIdentifiersByProjectRequest, SecretIdentifiersRequest, SecretsGetRequest},
+        ClientSecretsExt,
+    },
+    Client,
+};
+use color_eyre::eyre::{bail, Result};
+use uuid::Uuid;
+use which::which;
+
+use crate::{
+    util::{is_valid_posix_name, uuid_to_posix},
+    ACCESS_TOKEN_KEY_VAR_NAME,
+};
+
+pub(crate) async fn run(
+    client: Client,
+    organization_id: Uuid,
+    project_id: Option<Uuid>,
+    uuids_as_keynames: bool,
+    no_inherit_env: bool,
+    shell: Option<String>,
+    command: Vec<String>,
+) -> Result<i32> {
+    let is_windows = std::env::consts::OS == "windows";
+
+    let shell = match is_windows {
+        true => shell.unwrap_or_else(|| "powershell".to_string()),
+        false => shell.unwrap_or_else(|| "sh".to_string()),
+    };
+
+    if which(&shell).is_err() {
+        bail!("Shell '{}' not found", shell);
+    }
+
+    let user_command = if command.is_empty() {
+        if std::io::stdin().is_terminal() {
+            bail!("No command provided");
+        }
+
+        let mut buffer = String::new();
+        std::io::stdin().read_to_string(&mut buffer)?;
+        buffer
+    } else {
+        command.join(" ")
+    };
+
+    let res = if let Some(project_id) = project_id {
+        client
+            .secrets()
+            .list_by_project(&SecretIdentifiersByProjectRequest { project_id })
+            .await?
+    } else {
+        client
+            .secrets()
+            .list(&SecretIdentifiersRequest { organization_id })
+            .await?
+    };
+
+    let secret_ids = res.data.into_iter().map(|e| e.id).collect();
+    let secrets = client
+        .secrets()
+        .get_by_ids(SecretsGetRequest { ids: secret_ids })
+        .await?
+        .data;
+
+    if !uuids_as_keynames {
+        let mut seen = HashSet::new();
+        if let Some(s) = secrets.iter().find(|s| !seen.insert(&s.key)) {
+            bail!("Multiple secrets with name: '{}'. Use --uuids-as-keynames or use unique names for secrets", s.key);
+        }
```

Comment:
suggestion: We should either extract a helper function here or use itertools. Itertools is already a transient dependency for all our crates, and an explicit in some.

```suggestion
        if let Some(duplicate) = secrets.iter().map(|s| &s.key).duplicates().next() {
            bail!("Multiple secrets with name: '{}'. Use --uuids-as-keynames or use unique names for secrets", duplicate);
        }
```

Reply:
Applied in b86d0265ecfef374e2cb6042472f13b0688342f5 and added the dependency in f9456b8b0879564350b81e2d3c6147fc921a40d6.